vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js

```diff
@@ -78,69 +132,176 @@
+                        let dayMatch, monthMatch, yearMatch;
+
                         for (let i = 1; i < 4; i++) {
-                            datepicker.$connector.parts[i-1].value = parseInt(match[i]);
+                            const part = parts[i-1];
+                            const matchValue = match[i];
+                            part.value = parseInt(matchValue);
+                            if (part === dayPart) {
+                                day = part.value;
+                                dayMatch = matchValue;
+                            } else if (part === monthPart) {
+                                month = part.value - 1;
+                                monthMatch = matchValue;
+                            } else if (part === yearPart) {
+                                year = part.value;
+                                yearMatch = matchValue;
+                            }
                         }
-                        return {
-                            day: datepicker.$connector.dayPart.value,
-                            month: datepicker.$connector.monthPart.value - 1,
-                            year: datepicker.$connector.yearPart.value
-                        };
-                    }  else {
+                        return {day, month, year, dayMatch, monthMatch, yearMatch};
+                    } else {
                         return false;
                     }
-                });
+                };
+                _parseDate = tryCatchWrapper(_parseDate);
+
+                /**
+                 * @param {string} dateString
+                 * @returns {DateHash|boolean|undefined}
+                 */
+                const parseDate = function (dateString) {
+                    const date = _parseDate(dateString);
+                    if (!date) {
+                        return date;
+                    }
+                    if (date.yearMatch.length < 3 && date.year >= 0) {
+                        date.year += date.year < 50 ? 2000 : 1900;
+                    }
+                    return date;
+                };
+                datepicker.i18n.parseDate = tryCatchWrapper(parseDate);
 
                 if (inputValue === "") {
                     oldLocale = locale;
                 } else if (currentDate) {
                     /* set current date to invoke use of new locale */
-                    datepicker._selectedDate = new Date(currentDate.year, currentDate.month, currentDate.day);
+                    const rawDate = new Date(currentDate.year, currentDate.month, currentDate.day);
+                    if (currentDate.year >= 0 && currentDate.year <= 99) {
+                        // JS Date constructor converts years 0-99 to 1900-1999 so
+                        // this is needed to fix that when we want an explicit year.
+                        rawDate.setFullYear(currentDate.year);
+                    }
+                    datepicker._selectedDate = rawDate;
```

Comment:
Same fix as above in `formatDate()`. Without this fix a currently selected date with year less than 100 was not correctly kept when switching locale.

Reply:
1. `JS Date constructor` converts automatically the year from 50-99 to 1950-1999 already, and 0-49 to 2000-2049;

2. if `currentDate.year <= 99`=> the `JS Date constructor` converts automaticlly the year from 50-99 to 1950-1999 already, and 0-49 to 2000-2049;
 so do we need: to check `if (currentDate.year >= 0 && currentDate.year <= 99)` and  setFullYear`rawDate.setFullYear(currentDate.year);`- the year is set already by the constructor? 
It is just my opinion.

Reply:
Thought wrong your idea, so  I removed my comment.